packages/react-pages/src/node/dynamic-modules/PageStrategy.ts

```diff
@@ -0,0 +1,297 @@
+import * as fs from 'fs-extra'
+import * as path from 'path'
+import { EventEmitter } from 'events'
+import chokidar, { FSWatcher } from 'chokidar'
+import { debounce } from 'mini-debounce'
+import { PagesData } from './pages'
+import {
+  extractStaticData,
+  defaultPageLoader,
+  defaultPageFinder,
+} from './utils'
+
+export class PageStrategy extends EventEmitter {
+  readonly getPages: () => Promise<Readonly<PagesData>>
+  readonly close: () => void
+
+  constructor(
+    pagesDir: string,
+    findPages: FindPages = defaultPageFinder,
+    loadPageData: LoadPageData = defaultPageLoader
+  ) {
+    super()
+
+    const pageCache: PagesData = {}
+    const fileCache: FileCache = {}
+    const watchers = new Set<FSWatcher>()
+
+    // The "promise" event is for HMR updates to the pages map.
+    const emitPromise = debounce(() => {
+      this.emit('promise', pagesPromise)
+    }, 100)
+
+    // The "change" event is for HMR updates to each page.
+    const changedPages = new Set<string>()
+    const emitChange = debounce(() => {
+      this.emit('change', Array.from(changedPages))
+      changedPages.clear()
+    }, 100)
+
+    const helpers: PageHelpers = {
+      extractStaticData,
+      loadPageData: (file) => Promise.resolve(loadPageData(file, helpers)),
+      addPageData,
+      watchFiles,
+    }
+
+    // The file currently being processed.
+    let currentFile: File | null = null
+
+    let pagesPromise = Promise.resolve(pageCache)
+    findPages(pagesDir, helpers)
+
+    this.getPages = () => pagesPromise
+    this.close = () => watchers.forEach((w) => w.close())
+
+    function addPageData({
+      pageId,
+      key = 'main',
+      dataPath,
+      staticData,
+    }: PageData) {
+      if (!pageId.startsWith('/'))
+        throw Error(
+          `addPageData error: pageId should start with "/", but got "${pageId}"`
+        )
+
+      const existingPage = pageCache[pageId]
+      const page =
+        existingPage ||
+        (pageCache[pageId] = {
+          data: {},
+          staticData: {},
+        })
+
+      if (dataPath) {
+        if (page.data[key])
+          throw Error(
+            `addPageData conflict: data with key "${key}" already exists`
+          )
+        page.data[key] = dataPath
+      }
+
+      if (staticData) {
+        if (page.staticData[key])
+          throw Error(
+            `addPageData conflict: staticData with key "${key}" already exists`
+          )
+        page.staticData[key] = staticData
+      }
+
+      if (currentFile) {
```

Comment:
The `addPageData` could be called by users after the file is processed (.e.g calling `addPageData` upon some external event). So `currentFile` may be null or other file.

Why don't we find the file by dataPath? It is simpler and don't need global variable.

Reply:
Currently, calls to `addPageData` **must** be made within a `watchFiles` callback. Is there a use case for calling `addPageData` outside that? If nothing comes to mind, we should wait for users to request such capability.

> Why don't we find the file object by dataPath?

Because that's an optional property

Reply:
> Currently, calls to addPageData must be made within a watchFiles callback for cache invalidation to be automatic.

I can't understand... Could you give me an example? In the mental model I mentioned above, I don't think we need this restriction.

> Because that's an optional property.

If dataPath is undefined, then we don't need to do anything about file. So I think it is ok.

Reply:
> Currently, calls to addPageData must be made within a watchFiles callback for cache invalidation to be automatic.

Could you explain this to me? I still think the `currentFile` restriction is not needed. As an example, vite need to associate one file with multiple module objects. But vite seems not having such restriction.

Reply:
> Mostly about less global variable and code readability...

It's not a global variable. It's implicit context, and it's the only **reliable** way to associate `addPageData` calls with a specific file, so I see nothing wrong with it. For example, React hooks are implemented in much the same way.

> If dataPath is undefined, then we don't need to do anything about file.

That would lead to memory leaks. You would need a `removePageData` helper to prevent that, or `addPageData` calls without `dataPath` would be restricted to `watchFiles` callback.

In the end, I don't see the problem with restricting all `addPageData` calls to a `watchFiles` callback. It's not a big deal.